alea/include/alps/alea/autocorr.hpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 1998-2017 ALPS Collaboration. See COPYRIGHT.TXT
+ * All rights reserved. Use is subject to license terms. See LICENSE.TXT
+ * For use in publications, see ACKNOWLEDGE.TXT
+ */
+#pragma once
+
+#include <alps/alea/core.hpp>
+#include <alps/alea/util.hpp>
+#include <alps/alea/variance.hpp>
+
+#include <vector>
+
+// Forward declarations
+
+namespace alps { namespace alea {
+    template <typename T> class autocorr_acc;
+    template <typename T> class autocorr_result;
+}}
+
+// Actual declarations
+
+namespace alps { namespace alea {
+
+/**
+ * Accumulator for the integrated autocorrelation time.
+ *
+ * The integrated autocorrelation time `tau_int` of a time series is defined
+ * as the large-n limit of:
+ *
+ *                   1 + 2 * tau_int = n * var(n) / var(1),                (A)
+ *
+ * where `var(n)` is the sample variance obtained when averaging over batches
+ * batches, each batch being the mean of `n` consecutive elements of the series.
+ * Given a simulation of `N` steps, its corresponding squared error `sq_error`
+ * must thus be corrected as:
+ *
+ *                sq_error = (1 + 2 * tau_int) * var(1) / N                (B)
+ *
+ * which can be seen as replacing `N` with the number of uncorrelated samples.
+ * For a finite simulation, a tradeoff must be made between
+ *
+ *   (1) formal validity of above equations, which improves with `n`,
+ *   (2) statistical uncertainty in `tau_int`, which improves with `N/n`.
+ *
+ * This can be seen by plugging (A) into (B), which just yields the normal
+ * error estimate when sampling over bins of size `n`.
+ *
+ * The class builds up a hierarchy of variance estimates for different batch
+ * sizes, starting with `n=batch_size` at level 0, and increasing by a factor
+ * `granularity` at each level. Assuming `k`-sized vectors, the estimator
+ * scales as `O(k * log N)` in memory and `O(k * N * log log N)` in runtime.
+ */
+template <typename T>
+class autocorr_acc
+{
+public:
+    typedef T value_type;
+    typedef typename bind<circular_var, T>::var_type var_type;
+    typedef var_acc<T, circular_var> level_acc_type;
+
+public:
+    autocorr_acc();
+
+    autocorr_acc(size_t size=0, size_t batch_size=1, size_t granularity=2);
+
+    void reset();
+
+    bool initialized() const { return size_ != (size_t)-1; }
+
+    bool valid() const { return !level_.empty(); }
```

Comment:
What is the major difference between valid and initialized?

Reply:
See comment on that below.

This seems to be a point of confusion, maybe we should scrap the uninitialized state alltogether